src/client/mod.rs

```diff
@@ -157,78 +158,55 @@ impl APIClient {
-                        let current = items.pop().expect("items should not be empty"); // We know items is not empty so this is safe
-                        return Some((current, (resp, buff, items)));
-                    }
-
-                    // Loop pulling in chunks of the response until we have
-                    // actionable values (something containing a newline)
-                    loop {
-                        match resp.chunk().await {
-                            Ok(Some(chunk)) => {
-                                // append it to our current buffer
-                                buff.extend_from_slice(&chunk);
-
-                                // If it contains a new line, we have a full item somewhere in our buffer
-                                // Note: It seems that the kube api escapes all new line characters excepting the ones that
-                                // separate distinct events.
-                                if chunk.contains(&b'\n') {
-                                    // Create a new buffer for our next iteration
-                                    // to move any partial lines to
-                                    // We an safely reuse the items vec as we know it's empty here.
-                                    let mut new_buff = Vec::new();
-
-                                    for line in buff.split(|x| x == &b'\n') {
-                                        match serde_json::from_slice(&line) {
-                                            Ok(val) => items.push(Ok(val)),
-                                            Err(e) if e.is_eof() => {
-                                                // If this is an EOF error then we do not have a complete
-                                                // kube object, so push the current line into our new buffer
-                                                // and add the newline character we stripped as part of the iterator.
-                                                new_buff.extend_from_slice(&line);
-                                                new_buff.push(b'\n');
-                                            }
-                                            Err(e) => items.push(Err(Error::SerdeError(e))),
+        let stream = futures::stream::unfold((res, Vec::new()), |(mut resp, mut buff)| {
+            async {
+                loop {
+                    match resp.chunk().await {
+                        Ok(Some(chunk)) => {
+                            buff.extend_from_slice(&chunk);
+
+                            // If we've encountered a newline, see if we have any items to yield
+                            if chunk.contains(&b'\n') {
+                                let mut new_buff = Vec::new();
+                                let mut items = Vec::new();
+
+                                // Split on newlines
+                                for line in buff.split(|x| x == &b'\n') {
```

Comment:
maybe `for line in buff.lines()` if that works - would also handle stray `\r` characters if they exist.